datafusion/common/src/tree_node.rs

```diff
@@ -33,6 +33,9 @@ use crate::Result;
 /// [`LogicalPlan`]: https://docs.rs/datafusion-expr/latest/datafusion_expr/logical_plan/enum.LogicalPlan.html
 /// [`Expr`]: https://docs.rs/datafusion-expr/latest/datafusion_expr/expr/enum.Expr.html
 pub trait TreeNode: Sized {
+    /// Returns all children of the TreeNode
+    fn children_nodes(&self) -> Vec<Self>;
```

Comment:
First try with `Vec<&Self>`: https://github.com/apache/arrow-datafusion/compare/main...viirya:arrow-datafusion:refactor_treenode?expand=1.

One trouble is the `TreeNode` implementation for `Arc<T>`:

```
impl<T: DynTreeNode + ?Sized> TreeNode for Arc<T> {
    fn children_nodes(&self) -> Vec<&Arc<T>> {
        // DynTreeNode.arc_children returns Vec<Arc<Self>> and this function cannot return reference of temporary object
        unimplemented!("Call arc_children instead")
    }
```